app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/upscan/UpscanController.scala

```diff
@@ -83,56 +84,101 @@ class UpscanController @Inject() (
+        multipart.dataParts
+          .get("reference")
+          .flatMap(_.headOption)
+          .fold(Future.successful(BadRequest("missing upscan file descriptor id"))) { reference =>
+            upscanService
+              .getUpscanFileDescriptor(subscribed.subscribedDetails.cgtReference, UpscanInitiateReference(reference))
+              .value
+              .flatMap {
+                case Right(maybeUpscanFileDescriptor) => {
+                  maybeUpscanFileDescriptor match {
+                    case Some(upscanFileDescriptor) => {
+                      val userFile =
+                        multipart.files
+                          .map(file => file.copy(ref = FileIO.fromPath(file.ref.path): Source[ByteString, Any]))
+                      val prepared: MultipartFormData[Source[ByteString, Any]] =
+                        multipart
+                          .copy(
+                            files = userFile,
+                            dataParts = upscanFileDescriptor.fileDescriptor.uploadRequest.fields
+                              .mapValues(fieldValue => Seq(fieldValue))
+                          )
+                      upscanConnector
+                        .upload(upscanFileDescriptor.fileDescriptor.uploadRequest.href, prepared)
+                        .value
+                        .flatMap {
+                          case Left(error) =>
+                            logger.warn(s"could not upload file to S3", error)
+                            Future.successful(errorHandler.errorResult(None))
+                          case Right(_) => {
+                            upscanConnector
+                              .updateUpscanFileDescriptorStatus(upscanFileDescriptor.copy(status = UPLOADED))
+                              .value
+                              .map {
+                                case Left(error) => {
+                                  logger.warn(s"could not update upscan initiate file descriptor upload status: $error")
+                                  InternalServerError
+                                }
+                                case Right(_) => Ok
+                              }
+                          }
+                        }
+                    }
+                    case None => {
+                      logger.warn(
+                        s"could not find upscan initiate file descriptors with parameters: " +
+                          s"CGT Reference: ${subscribed.subscribedDetails.cgtReference.value} | Upscan Reference: $reference"
+                      )
+                      BadRequest
+                    }
+                  }
+                }
+                case Left(e) => {
+                  logger.warn(s"failed to get upscan file descriptor: $e")
+                  BadRequest
+                }
+              }
           }
       }
     }
-  }
+
+  def callBack(cgtReference: String): Action[JsValue] =
```

Comment:
I'd check to see if this can be done in the BE

Reply:
moved to BE